### ISSUE TITLE
Have not implemented kind name yet.

### ISSUE THREAD

**Issue by leighman** [prettier/plugin-php#1320]
Every so often the plugin seems to get upset and replace every function name and constant with the string 'Have not implemented kind name yet.'.
This happened once before and I was able to resolve by reinstalling but can't seem to this time.
Looks like two bugs:
1. Something going wrong and producing a node that the printer doesn't understand
2. Error from the printer making it into the output rather than prettier stopping.
Happy to look into this if you can provide pointers for how I might start.

Prettier: 1.19.1
Plugin: 0.12.0

**Comment by czosel** [comment]
Hi @leighman,
Interesting, I didn’t see this yet. Can you reproduce the issue, i.e. share some code that shows this behavior?

**Comment by leighman** [comment]
Seems to be something to do with project configuration.
Doesn't happen in a new project and happens in even the simplest file within this project.
Any pointers for debugging?

**Comment by czosel** [comment]
Can you share your .prettierrc?

**Comment by leighman** [comment]
```
{
  "trailingComma": "es5",
  "semi": false,
  "singleQuote": true
}
```
We don't have any specific php settings.

I think possibly what was happening is `yarn install` was bailing because `fsevents` failed to build and that meant that one of the libraries the plugin relied on was not getting installed. Does that sound possible? So broken on a fresh install but working on reinstall.

**Comment by czosel** [comment]
I could imagine, yeah. We only have three runtime dependencies next to prettier core, so you could check if those are correctly installed. Another possible cause might be that due to the failed install you’re actually running an older version of the plugin?

**Comment by loilo** [comment]
I can consistently reproduce this in the current 0.13 prerelease. However, due to the breaking changes in the API, I cannot provide a playground link yet.

The cause is an unrecognized node kind (originating [here](https://github.com/prettier/plugin-php/blob/da68fd6fa1b2c6717eeb7e5be15ebaddc4ac0154/src/printer.js#L2868)) when the formatter tries to handle a `node.what` object as a regular node.

Sample code:

```php
<?php

array_map();
```

You can see in the AST below, where the `"kind": "classreference"` happens.

```json
{
  "kind": "program",
  "loc": {
    "source": "<?php\n\narray_map();\n\n",
    "start": {
      "line": 1,
      "column": 0,
      "offset": 0
    },
    "end": {
      "line": 5,
      "column": 0,
      "offset": 21
    }
  },
  "children": [
    {
      "kind": "expressionstatement",
      "loc": {
        "source": "array_map();",
        "start": {
          "line": 3,
          "column": 0,
          "offset": 7
        },
        "end": {
          "line": 3,
          "column": 12,
          "offset": 19
        }
      },
      "expression": {
        "kind": "call",
        "loc": {
          "source": "array_map();",
          "start": {
            "line": 3,
            "column": 0,
            "offset": 7
          },
          "end": {
            "line": 3,
            "column": 12,
            "offset": 19
          }
        },
        "what": {
          "kind": "classreference",
          "loc": {
            "source": "array_map",
            "start": {
              "line": 3,
              "column": 0,
              "offset": 7
            },
            "end": {
              "line": 3,
              "column": 9,
              "offset": 16
            }
          },
          "resolution": "uqn",
          "name": "array_map"
        },
        "arguments": []
      }
    }
  ],
  "errors": [],
  "comments": [],
  "extra": {
    "parseAsEval": false
  }
}
```

**Comment by czosel** [comment]
Oh, that’s not good - I’ll try to take a look tomorrow. Note that 0.13.0 contains no breaking API changes (see corresponding release branch).

**Comment by loilo** [comment]
Oh, okay — sorry, I just assumed that 0.13 was created from master.

Then let me rephrase this:

I **can** reproduce the issue from the current master HEAD commit.

I **can not** reproduce this issue with the 0.13 build.

**Comment by czosel** [comment]
Ah good, I was afraid I released something broken 😅 Still great to have a reproduction for this 👍

**Comment by loilo** [comment]
Nope, all good. Although I'd almost expect things to randomly break in a pre-release. 🙈

**Comment by loilo** [comment]
By the way, and in the lack of a better place for mentioning it: I've put up an alternative version of the playground [here](https://dazzling-kirch-442b3d.netlify.com/) with the new API structure enabled (`phpVersion` option, boolean `trailingCommaPhp`). You need to provide your own standalone build from master though.

**Comment by czosel** [comment]
I was just able to reproduce the issue locally: Most tests were failing with `Have not implemented kind classreference yet.`. After running `yarn cache clean && rm -rf node_modules && yarn` the issue went away. Can you try and confirm @loilo?

**Comment by loilo** [comment]
Can confirm. Went from 146 failed tests to 0. 🤷‍♂

So apparently a caching issue. Anything for us to do here?

**Comment by czosel** [comment]
Yeah, this happens because we’re referencing the parser by commit hash, and the yarn cache seems to handle them incorrectly sometimes.

I think there’s not much to be done, so I’ll go ahead and close the issue - @leighman, feel free to reopen it if you can reproduce the issue with a fresh install!

**Comment by leighman** [comment]
`yarn` seems to struggle between
```
-php-parser@glayzzle/php-parser#4e9d584c7bd478175777e643e8d9d57d1863023a:
+"php-parser@github:glayzzle/php-parser#4e9d584c7bd478175777e643e8d9d57d1863023a":
```
in the `yarn.lock`. Could it be related to that?
With the above change (as a result of removing `node_modules` and `yarn.lock`) it seemed to be fixed for me.

**Comment by czosel** [comment]
Hmm, when I remove `yarn.lock` and re-install, I don't get any diff related to `php-parser` in the lockfile. What version of yarn are you using?

**Comment by leighman** [comment]
1.21.1

**Comment by czosel** [comment]
I don't get the diff with 1.22.0 - can you confirm?
